os/pm/pm_ioctl.c

```diff
@@ -0,0 +1,143 @@
+
+#include <tinyara/config.h>
+#include <assert.h>
+#include <tinyara/pm/pm.h>
+#include <tinyara/clock.h>
+#include <tinyara/irq.h>
+
+#include "pm.h"
+
+#ifdef CONFIG_PM
+
+/****************************************************************************
+ * Public Functions
+ ****************************************************************************/
+
+/****************************************************************************
+ * Name: pm_ioctl
+ *
+ * Description:
+ *   This function can be called from applications running on TizenRT to
+ *   request/enable/disable power state transitions. This function conveys
+ *   the request from the application to the PM module, which thereafter
+ *   checks the feasibility of the requests. There is no returned value.
+ *
+ *   The power management state is not automatically changed, however.
+ *   up_idlepm() must be called in order to make the state change.
+ *
+ *   These two steps are separated because the application is not supposed
+ *   to invoke power state changes directly, it can only request/suggest.
+ *
+ * Input Parameters:
+ *   domain - the PM domain to check
+ *
+ * Returned Value:
+ *   None
+ *
+ ****************************************************************************/
+
+void pm_ioctl(enum pm_ioctl_cmds cmd)
+{
+	switch(cmd) {
+		case PM_IOC_NOCMD:
+			dbg("Nothing to do\n");
+			break;
+		case PM_IOC_STAY:
+			//Get the current domain
+			//perform pm_stay to block the domain in its current PM state
+			//pm_stay(<DOMAIN>, pm_querystate(<DOMAIN>));
+			break;
+		case PM_IOC_RELAX:
+			//Get the current domain
+			//perform pm_relax to unblock the domain's state transitions
+			//pm_relax(<DOMAIN>, pm_querystate(<DOMAIN>));
+			break;
+		/* Note that in the above 2 IOCTL calls, the application does not provide the state to stay/relax.
+		 * The application cannot anyways know the pm_domain of the current thread.
+		 * The thread will be extracted during the processing of the IOCTL call itself.
+		 * The state to stay in is the current state itself, and the state to be relaxed from will also
+		 * be the current state itself. This is based on the assumption that if an application is requesting
+		 * a stay, then the current state is capable enough for the application to do its work, hence we will
+		 * stay in the current request only. If a stay was invoked in the current PM state, then it should be
+		 * relaxed from the same state too.
+		 * Open to discussion and consideration
+		 */
+		case PM_IOC_SLEEP:
+			//Get the current domain
+			/* Perform any checks neccessary for doing the next step. As per current code in idle.c, some
+			 * preliminary steps are performed in STANDBY state, and then state changes to SLEEP.
+			 */
+			/* Change the recommended state for the current domain to PM_SLEEP
+			 * pdom->recommended = PM_SLEEP
+			 */
+			/* up_idlepm() -> The code in this function may be needed to be modified a bit so that the opening
```

Comment:
`	/* Decide, which power saving level can be obtained */
	newstate = pm_checkstate(PM_IDLE_DOMAIN);

	/* Check for state changes */
	if (newstate != oldstate)
	{
`
This is the opening condition for up_idlepm(), so far I don't think there is any changes need to be applied here.
Could you elaborate more of your thoughts?

Besides, I think it is more important to take care of the stability of PM cycle flow which are the codes under os/pm for jumping the state directly to PM_SLEEP, please refer to the comment above.

Reply:
case PM_IOC_SLEEP:
in pm_ioctl() sets pdom->recommended to PM_SLEEP and calls up_idlepm()
Then in idlepm() we again perform pm_checkstate()
- firstly, I think that calling checkstate() is wasteful here because recommended state is already updated.
- second, since recommended state was changed to sleep in a non-natural way (I mean not through natural ladder transition), what is the chance that pdom->recommended again gets updated to some state other than PM_SLEEP in pm_checkstate() and hence the board does not go to sleep now.

Reply:
May I understand it as, other than the infinite loop for up_idlepm() handled in os_start.c, you want to generalize it so that the pm_ioctl can also invoke up_idlepm() for state transition directly to PM_SLEEP?

Reply:
Yes, that is my idea of creating the link for application to be able to invoke sleep. You may choose a different kind of implementation too.

Reply:
In my opinion, I think it might be a good approach to skip through unnecessary state transitions if the application is ready, since most of the driver check is being done in the PM_SLEEP state.
By the way, from our last meeting, I think I have shared that the current structure for PM logic will draw the power consumption higher than applying asm("wfi"), is there any consideration to apply such modes to compromise with such condition? Or you think it is okay to view the power consumption for current PM structure as the baseline? ie. ~65mA compared to ~25mA

Reply:
1. Do you mean completely skipping the ladder mechanism?
2. If yes, do you suggest that always or only in the case when App wants to invoke sleep?
3. We can also reduce the time halt in NORMAL, IDLE and STANDBY states to move to sleep quicker
4. Also, I think we should keep the ladder mechanism in idle_thread i.e. when idle loop transitions to SLEEP state. However, when app requests sleep, we can skip the full ladder transition as it takes time. We can assume that since app has requested sleep,  that means app is intelligently aware of how much downtime is coming up next for the CPU.

Regarding the current/power consumption, I will share in a separate communication.

Reply:
1 & 2 )The ladder like mechanism should be maintained as it is, only application should have the right to invoke direct transition to sleep, but that requires some more test to verify our approach is feasible with current PM driver.
3) Do you have any plan to make this timer gap dynamically adjustable, or changing in defconfig is enough?
4) Agree.

Reply:
^3) The gap may be adjusted dynamically during runtime, depending on how useful that proves to be. In the least, the timer values can definitely be updated to the most optimal values instead of the current ones.
Either of the above 2 approaches would require data analysis from practical/pseudo-practical scenarios, and I think that would overlap with the current consumption problem.

Reply:
Understood, it seems statically assignment is enough for the operation, and I feel the discussion timing for this might be too early. Lets bring this timer gap topic back at later stage.